apps/s_client.c

```diff
@@ -3025,15 +3025,22 @@ int s_client_main(int argc, char **argv)
     if (in_init)
         print_stuff(bio_c_out, con, full_log);
     do_ssl_shutdown(con);
-#if defined(OPENSSL_SYS_WINDOWS)
     /*
-     * Give the socket time to send its last data before we close it.
-     * No amount of setting SO_LINGER etc on the socket seems to persuade
-     * Windows to send the data before closing the socket...but sleeping
-     * for a short time seems to do it (units in ms)
-     * TODO: Find a better way to do this
+     * If we ended with an alert being sent, but still with data in the
+     * network buffer to be read, then calling BIO_closesocket() will
+     * result in a TCP-RST being sent. On some platforms (notably
+     * Windows) then this will result in the peer immediately abandoning
+     * the connection including any buffered alert data before it has
+     * had a chance to be read. Shutting down the sending side first,
+     * and then closing the socket sends TCP-FIN first followed by
+     * TCP-RST. This seems to allow the peer to read the alert data.
      */
-    Sleep(50);
```

Comment:
This sleep, together with its comment, needs to remain.  I'm now running a test loop with this PR and this Sleep(50) added back, and the test loop is now almost at a whooping 200 rounds!  (I haven't seen it go past 70 before)

Reply:
Thanks for testing.

This patch also fixes your proxy patch, it has now been running for more than 1700 times .

Reply:
Are the defines _WIN32 and SD_SEND available in the cygwin build?

Reply:
to answer my question, that must be the case, otherwise it would not have made any difference...

Reply:
Silly me, I forget that Cygwin is its own beast, so I'm currently trying with this:

``` C
#if defined(OPENSSL_SYS_WINDOWS)
    Sleep(50);
#elif defined(OPENSSL_SYS_CYGWIN)
    usleep(50000);
#endif
```

Reply:
Cygwin is aiming to be a POSIX library / environment, so I think we need to think accordingly, even though there's the Windows API somewhere far behind

Reply:
Re `shutdown`, I usually go with the numeric arguments, they are the same everywhere.

Reply:
yeah, looks like you should make an own PR for it.

Reply:
> Re shutdown, I usually go with the numeric arguments, they are the same everywhere.

You know, not everyone would agree with that.
At least for readability of the code using named defines is a win.